Compiler/src/typeinfer.jl

```diff
@@ -148,7 +148,10 @@ function finish!(interp::AbstractInterpreter, caller::InferenceState, validation
             result.analysis_results, time_total, caller.time_caches, time_self_ns * 1e-9, di, edges)
         engine_reject(interp, ci)
         codegen = codegen_cache(interp)
-        if !discard_src && codegen !== nothing && uncompressed isa CodeInfo
+        if !discard_src && codegen !== nothing && uncompressed !== nothing
+            if !(uncompressed isa CodeInfo)
+                uncompressed = ir_to_codeinf!(uncompressed)
+            end
```

Comment:
I wasn't sure why this part needed to be changed. Perhaps we could use `inferred_result` here instead of `uncompressed`? That way, if we properly define `transform_result_for_cache`, it would also provide support for `codegen_cache`.

Reply:
It is also now missing a check for isa CodeInfo, which may cause problems for other consumers